### PR TITLE
[CI] Add env var suggestion for common call ray start fixture error on MacOS

### LINKED ISSUE #35229: [Core] [Devprod] `call_ray_start` fixture returns wrong address, causing tests to fail when run locally on Mac OS
### What happened + What you expected to happen

In a lot of tests we use the pattern `address = call_ray_start; ray.init(address)`.  This causes tests to fail when running tests locally:
```
RuntimeError: Module: s collection is enabled. To disable this, add `--disable-usage-stats` to the command that starts the cluster, or run the following command: `ray disable-usage-stats` before starting the cluster. See https does not exist.
```

The reason is that `call_ray_start` returns the following:
```python
    # Get the redis address from the output.
    redis_substring_prefix = "--address='"
    address_location = out.find(redis_substring_prefix) + len(redis_substring_prefix)
    address = out[address_location:]
    address = address.split("'")[0]

    yield address
```

Here `out` is the output of `ray start --head`.  But on Mac, the string `--address` appears nowhere in the output.  The root cause is that `ray_constants.ENABLE_RAY_CLUSTER` is not set on Mac, because multinode clusters are not supported on Mac:

https://github.com/ray-project/ray/blob/0006a81abb7b44480263580591fd7acb17a6fde3/python/ray/scripts/scripts.py#L761-L771

  The workaround is to set the environment variable `RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER=1`, but the test fixture should handle this more gracefully. The error can be hard to debug.

### Versions / Dependencies

Ray master, Mac OS (and likely on windows)

### Reproduction script

In any pytest file:
```
@pytest.mark.asyncio
@pytest.mark.parametrize(
    "call_ray_start",
    ["ray start --head"],
    indirect=True,
)
async def test_my_test(call_ray_start):
    address_info = ray.init(address=call_ray_start)
```

### Issue Severity

Low: It annoys or frustrates me.

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
The `call_ray_start` fixture was not correctly returning the GCS address when run locally on MacOS, causing any tests with that fixture to fail.  

The reason is that the address is parsed from the `ray start --head` output, but the "`To add another node to this Ray cluster, run ray start --address=<GCS address>`" line is not printed on MacOS because multi-node clusters are disabled by default on MacOS.

To fix this, you need to enable multinode clusters before running the test.  This environment variable is set in Buildkite CI, but Mac developers might not have it set on their local machine.  This PR prints a suggestion to set the environment variable when this failure is encountered.

## Related issue number
Closes https://github.com/ray-project/ray/issues/35229

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/conftest.py [hunk 1]
```diff
@@ -463,7 +463,11 @@ def ray_start_object_store_memory(request, maybe_external_redis):
 
 
 @pytest.fixture
-def call_ray_start(request):
+def call_ray_start(request, monkeypatch):
+    # Force "--address=<GCS address>" to appear in the output so we can parse
+    # the address.
+    monkeypatch.setenv(ray_constants.ENABLE_RAY_CLUSTERS_ENV_VAR, "1")
```
Comment (wuisawesome) [inline python/ray/tests/conftest.py:469]: Can we set the env var on line 482 instead of introducing a monkey patch here?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
let's avoid monkeypatching if possible

**Comment by architkulkarni** [comment]
@wuisawesome It's not a monkeypatch, it actually sets the env var for the lifetime of the fixture.  It's a pytest fixture that makes sure the global env is untouched.

**Comment by architkulkarni** [comment]
Or do you mean to set it in the command only? like `ENABLE_CLUSTERS=1 ray start ...`

**Comment by jjyao** [comment]
We have `    - export RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER=1` in pipeline.macos.yaml, is it not working?

**Comment by architkulkarni** [comment]
@jjyao It is working in Buildkite CI.  This PR only affects developers on Mac who are running tests locally.  Currently the error message is hard to debug.

**Comment by jjyao** [comment]
Oh, I see. If you run the test locally via bazel, it should have the env var set. We have `test --test_env=RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER="1"` in `.bazelrc`.

**Comment by architkulkarni** [comment]
Sure.  I and others use `pytest -vs test_job_manager.py::my_specific_unit_test`, not sure if there's a convenient way to do that in bazel.

**Comment by jjyao** [comment]
Yea, something like `bazel test //python/ray/tests:test_ray_init`. I think this is the preferred way since we used bazel in CI as well. If we use pytest directly then we lose whatever inside `.bazelrc`

**Comment by architkulkarni** [comment]
Oh but how do you run a specific function or class within the `.py` test file?  

If bazel is the best way, I can update https://docs.ray.io/en/latest/ray-contribute/getting-involved.html#testing-for-python-development and get rid of the references to pytest

**Comment by jjyao** [comment]
For that you can do `bazel test //python/ray/tests:test_ray_init --test_arg=-k "test_xxx"` but we also need to change the main function of the test script to 
```
if __name__ == "__main__":
    import pytest
    sys.exit(pytest.main(["-v", __file__] + sys.argv[1:]))
```

in order to pass through the `-k` stuff.

**Comment by wuisawesome** [comment]
I'm saying `ray_context` has code like this 

```
    env = None
    if isinstance(parameter, dict):
        if "env" in parameter:
            env = {**os.environ, **parameter.get("env")}
```

Shouldn't we just make that 

    env = {"RAY_ENABLE_WINDOWS_OR_OSX_CLUSTER": 1}
    if isinstance(parameter, dict):
        if "env" in parameter:
            env = {**os.environ, **parameter.get("env")}.update(env)

**Comment by architkulkarni** [comment]
Updated the PR to just print the warning.  It's a small one time cost for developers to enable the env var locally if they want to continue to use `pytest` directly with all of its CLI features instead of using `bazel`. This way we also don't have to worry about the CI env set in Buildkite being overwritten in a random fixture (could see it being hard to debug in the future).

**Review by wuisawesome — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]

**Comment by architkulkarni** [comment]
`//python/ray/tune:test_controller_resources_integration                  FAILED in 3 out of 3 in 315.5s`
failure unrelated

Windows `test_scheduling*`, `test_basic_4*`, `test_task_events` unrelated

**Comment by vitsai** [comment]
Unfortunately windows://python/ray/tests:test_basic_4_client_mode and windows://python/ray/tests:test_basic_4 both started failing with this commit, so reverting for now.

**Pull request merged**